Callable arguments of methods/functions are checked if they are valid callbacks

MartinMystikJonas opened the pull request.

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,124 @@
+	public function checkCallableArgument(
+		Arg $argument,
+		Scope $scope,
+		string $msgPrefix
```
Please don't use abbreviations.

ondrejmirtes replied:
Please don't use abbreviations.

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,124 @@
+				return [$msgPrefix . "array doesn't have required 2 items."];
+			}
+			if ($arrayItemsCount > 2) {
+				return [$msgPrefix . 'array has too much items, only two items expected.'];
```
has too many

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,124 @@
+				$targetType = new ObjectType($item1->value);
+				$targetValue = $item1->value;
+				$targetIsStaticMethod = true;
+			} elseif ($item1 instanceof Variable) {
```
Does not have to be variable, e.g. `[$this->object, 'method']`. This should be the fallback `else` branch.

ondrejmirtes replied:
Does not have to be variable, e.g. `[$this->object, 'method']`. This should be the fallback `else` branch.

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,124 @@
+				$targetType = $scope->getType($item1);
+				$targetIsStaticMethod = false;
+			} elseif ($item1 instanceof ConstFetch) {
+				$targetType = new ObjectType(constant($item1->name->toString()));
```
This will crash in case the constant does not exist.

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,124 @@
+				$targetType = new ObjectType(constant($item1->name->toString()));
+				$targetValue = constant($item1->name->toString());
+				$targetIsStaticMethod = true;
+			} elseif ($item1 instanceof Class_) {
```
This does not make sense. You probably meant `ClassConstFetch`.

ondrejmirtes replied:
Also think about `self::class`, `static::class`, `parent::class`...

MartinMystikJonas replied:
This should catch __CLASS__ (PhpParser\Node\Scalar\MagicConst\Class_) and it works in tests.

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,124 @@
+			}
+		}
+
+		if ($targetType === null && strpos($targetMethod, '::') !== false) {
```
This should have some sanity check so that it does not catch `::` at the beginning nor at the end of the string.

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,124 @@
+		if ($targetType !== null) {
+			// method
+			if (!$targetType->canCallMethods()) {
+				return [$msgPrefix . sprintf('array first item {%s}(%s) is not object with callable methods.', $targetType->describe(), $targetValue)];
```
We usually add an error message `Cannot call method %s() on %s.`, see for example CallMethodsRule.

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,124 @@
+			try {
+				$this->broker->getFunction(new Name($targetMethod), $scope);
+			} catch (\PHPStan\Broker\FunctionNotFoundException $e) {
+				return [$msgPrefix . sprintf('%s is not a existing function name.', $targetMethod)];
```
`Function %s not found.` is enough.

ondrejmirtes commented on src/Rules/Functions/FunctionCallableArgExistsRule.php:
```diff
@@ -0,0 +1,93 @@
+				continue;
+			}
+
+			$msgPrefix = sprintf('Argument #%d of %s should be callable, but passed ', $i + 1, $node->name);
```
The name of the argument is also added, for example: `'Parameter #%d %s of method ' . $messagesMethodName . ' expects %s, %s given.',` (see usages of FunctionCallParametersCheck).

ondrejmirtes commented on src/Rules/Methods/MethodCallableArgExistsRule.php:
```diff
@@ -0,0 +1,75 @@
+		$typeResult = $this->ruleLevelHelper->findTypeToCheck($scope, $node->var, '');
+		$type = $typeResult->getType();
+		if ($type instanceof ErrorType) {
+			return [];
```
`return $typeResult->getUnknownClassErrors()`

ondrejmirtes commented on tests/PHPStan/Rules/Functions/FunctionCallableArgExistsRuleTest.php:
```diff
@@ -0,0 +1,211 @@
+		$lineOffset2 = 74;
+
+		$this->analyse([__DIR__ . '/../data/callable-exists.php'], [
+					[
```
The test is really weirdly formatted, please look at it on GitHub.

ondrejmirtes commented on tests/PHPStan/Rules/Functions/FunctionCallableArgExistsRuleTest.php:
```diff
@@ -0,0 +1,211 @@
+		$this->analyse([__DIR__ . '/../data/callable-exists.php'], [
+					[
+				'Argument #1 of call_user_func should be callable, but passed array has too much items, only two items expected.',
+				$lineOffset, //109
```
Please don't use this, write the literal line where the error occurs.

MartinMystikJonas replied:
I will change it in final version, I often add new test cases whoch changes line numbers so i create offset to fix tests more easily.

ondrejmirtes commented:
I see, please also add support for ::class (ClassConstFetch).

On Wed, 27 Dec 2017 at 13:09, Martin Mystik Jonáš <notifications@github.com>
wrote:

> *@MartinMystikJonas* commented on this pull request.
> ------------------------------
>
> In src/Rules/CallableExistsCheck.php
> <https://github.com/phpstan/phpstan/pull/716#discussion_r158805069>:
>
> > +			}
> +			$item1 = $argument->value->items[0]->value;
> +			$item2 = $argument->value->items[1]->value;
> +
> +			if ($item1 instanceof String_) {
> +				$targetType = new ObjectType($item1->value);
> +				$targetValue = $item1->value;
> +				$targetIsStaticMethod = true;
> +			} elseif ($item1 instanceof Variable) {
> +				$targetType = $scope->getType($item1);
> +				$targetIsStaticMethod = false;
> +			} elseif ($item1 instanceof ConstFetch) {
> +				$targetType = new ObjectType(constant($item1->name->toString()));
> +				$targetValue = constant($item1->name->toString());
> +				$targetIsStaticMethod = true;
> +			} elseif ($item1 instanceof Class_) {
>
> This should catch *CLASS* (PhpParser\Node\Scalar\MagicConst\Class_) and
> it works in tests.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/phpstan/phpstan/pull/716#discussion_r158805069>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAGZuBv9kKtZM3aMi_rIbNZQd_jeoRhWks5tEjNpgaJpZM4RNBbd>
> .
>
-- 

Ondřej Mirtes

MartinMystikJonas commented:
New version commited:
- updated error messages
- support for ::class
- support for properties and static properties in args
- support for expressions in args (only check types)
- fixed formatting
- expanded tests to cover more edge cases

MartinMystikJonas commented:
I'm working on new version which will support wider variety of parametr combinations (class constant as method name, ...) and I also try to add support for concatenated values like:

```php
  call_user_func(self::class . "::someMethod");
```

MartinMystikJonas commented:
Commited new version with refactored CallableExistsCheck nad better support for value resolving.

@ondrejmirtes Method resolveValue is candidate for moving to Scope class if it would be made more universal. Currenty in only supports strings but can be extended to other scalars. This method resolves value of Expr to scalar if possible (const, class const, literal, concatenation, ...) or return Type if not (variable, function call, operator, ...).

MartinMystikJonas commented:
@ondrejmirtes Any idea how to get list of all core PHP functions that accepts callable as argument?

iluuu1994 commented:
@MartinMystikJonas Stubs will be needed for that.

https://github.com/JetBrains/phpstorm-stubs

ondrejmirtes commented:
@MartinMystikJonas Yep, just list a few, add support for functions/methods with callable argument (userland ones) and it will automatically work once we add these stubs in some form or another. (Do not attempt to add the stubs here, that would be out of scope of this PR.)

MartinMystikJonas commented:
@ondrejmirtes Listed all main functions i can find. So for now this feature is complete from my side. Please review it and let me know is anything more is required. Then I will clean up code, squash commits and prepare final version.

MartinMystikJonas commented:
Should I prepare final version or shpuld I wait fir review? This afternoon could be last I could work on this for few months.

mavimo commented on src/Rules/Functions/FunctionCallableArgExistsRule.php:
```diff
@@ -0,0 +1,114 @@
+			['array_uintersect_assoc', 4],
+			['array_intersect_uassoc', 4],
+			['array_uintersect_uassoc', 4],
+			['array_diff_ukey', 4],
```
As I understand we are indicate that parameter 4 into function should be a callable, but php doc indicate that the callable should be the last item of the function, so it call be the 3, the 4 or the N-th element. Same issue on array_filter, array_udiff, ..
```php
$a = [1, 2, 3];
$b = [4, 5, 6];
$c = [7, 8, 9];
$d = [10, 11, 12];
$e = [13, 14, 15];
$res = array_udiff($a, $b, $c, $d, $e, function (int $first, int $second): int {
  return $first - $second;
});
```

http://php.net/array_diff_ukey

MartinMystikJonas replied:
Good point.

mavimo replied:
@MartinMystikJonas Maybe should consider to get last element of args and process it when needed?

ondrejmirtes commented on tests/PHPStan/Rules/Functions/FunctionCallableArgExistsRuleTest.php:
```diff
@@ -0,0 +1,325 @@
+				'Argument #1 callableParam of funcWithCallableParam should be callable, but passed non-static method CallableExists\Bar::knownMethod as static.',
+				$lineOffset2 + 64,
+			],
+						[
```
Wrong indentation.

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,186 @@
+				$targetIsStaticMethod = true;
+			} elseif ($item1 instanceof ObjectType || $item1 instanceof ThisType) {
+				$targetType = $item1;
+				$targetIsStaticMethod = false;
```
Did you consider the case where the static method is called on an object? See: https://3v4l.org/HQ1cp

MartinMystikJonas replied:
This will result in error:

ondrejmirtes commented on src/Rules/CallableExistsCheck.php:
```diff
@@ -0,0 +1,186 @@
+			if (is_string($item1)) {
+				$targetType = new ObjectType($item1);
+				$targetIsStaticMethod = true;
+			} elseif ($item1 instanceof ObjectType || $item1 instanceof ThisType) {
```
I feel like this should be a different condition. What about not having the `else` branch and just asking the `Type` about `hasMethod`/`getMethod`?

ondrejmirtes replied:
This will solve union/intersection types, StaticType...

MartinMystikJonas replied:
I'm not sure what you mean here. How do you think this conditionshould look like?

ondrejmirtes replied:
Like this:

```
if (is_string($item1)) {
   ...
} elseif ($item1 instanceof MixedType || $item1 instanceof StringType) {
   ...
} else {
   // fallback for all other types - make sure to call $item1->hasMethod() + getMethod somewhere lower in this file
}
```

ondrejmirtes commented:
Please note that I updated master to PHP 7.1, adding nullable and void typehints, so be careful when rebasing 😊 Run `vendor/bin/phing cs` to find out anything you might have missed.

ondrejmirtes commented:
This might require some rework and/or simplification in light of #828 😊 `isCallable` now works more universally, see for example [`ConstantStringType::isCallable()`](https://github.com/phpstan/phpstan/pull/828/commits/8bc1d1bb115dd634355b56048d59c787b3eb2e7f#diff-ae3e4587957a9ee520cc6dd025381da7R30).

Note: #828 will be merged soon.

ondrejmirtes commented:
Well, it's even more advanced than I suspected, #828 now fully recognizes arrays and strings as callables. It does not have such nice error messages as this pull request but gets the job done 😊 Don't close this yet, I'll go through it to find out whether we forgot something.

With function stubs of native functions that are also in the works, this feature will be completely covered 😊
